### PR TITLE
Allow Random Generators to be instantiated with device types

### PR DESCRIPTION
**Pull request by crtrott**
Cleans up specialization as well !!

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by crtrott** [comment]
Retest this please

**Comment by crtrott** [comment]
Retest this please

File: algorithms/src/Kokkos_Random.hpp [hunk 1]
```diff
@@ -537,6 +537,114 @@ struct rand<Generator, Kokkos::complex<double> > {
...
+  KOKKOS_FUNCTION
+  Random_XorShift1024_State() : state_(nullptr), stride_(1){};
+
+  template <class StateViewType>
+  KOKKOS_FUNCTION Random_XorShift1024_State(const StateViewType& v,
+                                            int state_idx)
+      : state_(&v(state_idx, 0)), stride_(v.stride_1()) {}
+
+  KOKKOS_FUNCTION
+  uint64_t& operator[](const int i) const { return state_[i * stride_]; }
```
Comment (dalg24) [inline algorithms/src/Kokkos_Random.hpp]: Hell no.  This breaks constant correctness.

**Pull request merged**